packages/react-components/react-combobox/src/ComboboxBase/useComboboxBaseSlots.ts

```diff
@@ -0,0 +1,138 @@
+import * as React from 'react';
+import { useMergedRefs } from '@fluentui/react-utilities';
+import { getDropdownActionFromKey, getIndexFromAction } from '../utils/dropdownKeyActions';
+import type { ComboboxState } from '../components/Combobox/Combobox.types';
+import type { DropdownState } from '../components/Dropdown/Dropdown.types';
+import type { ComboboxBaseProps } from './ComboboxBase.types';
+
+export const useComboboxBaseSlots = (props: ComboboxBaseProps, state: ComboboxState | DropdownState) => {
+  const { multiselect, placeholder } = props;
+  const {
+    activeOption,
+    getCount,
+    getIndexOfId,
+    getOptionAtIndex,
+    open,
+    selectOption,
+    setActiveOption,
+    setOpen,
+    value,
+  } = state;
+
+  // handle trigger focus/blur
+  const triggerRef = React.useRef<HTMLButtonElement | HTMLInputElement>(null);
+  const ignoreTriggerBlur = React.useRef(false);
+
+  /* Add listbox props and event handlers  */
+  state.listbox = {
+    multiselect,
+    tabIndex: undefined,
+    ...state.listbox,
+  };
+
+  /*
+   * Handle focus when clicking the listbox popup:
+   * 1. Move focus back to the button/input when the listbox is clicked (otherwise it goes to body)
+   * 2. Do not close the listbox on button/input blur when clicking into the listbox
+   */
+  const { onClick: onListboxClick, onMouseDown: onListboxMouseDown } = state.listbox;
+  state.listbox.onClick = (event: React.MouseEvent<HTMLDivElement>) => {
+    triggerRef.current?.focus();
+
+    onListboxClick?.(event);
+  };
+
+  state.listbox.onMouseDown = (event: React.MouseEvent<HTMLDivElement>) => {
+    ignoreTriggerBlur.current = true;
+
+    onListboxMouseDown?.(event);
+  };
+
+  /* Add trigger props and event handlers */
+  let triggerSlot;
+
+  if ((state as ComboboxState).input) {
+    (state as ComboboxState).input = {
+      'aria-expanded': open,
+      'aria-activedescendant': open ? activeOption?.id : undefined,
+      placeholder,
+      role: 'combobox',
+      type: 'text',
+      value,
+      ...(state as ComboboxState).input,
+      ref: useMergedRefs((state as ComboboxState).input.ref, triggerRef as React.RefObject<HTMLInputElement>),
+    };
+
+    triggerSlot = (state as ComboboxState).input;
+  } else {
+    (state as DropdownState).button = {
+      'aria-expanded': open,
+      'aria-activedescendant': open ? activeOption?.id : undefined,
+      children: value ? value : placeholder,
+      role: 'combobox',
+      type: 'button',
+      ...(state as DropdownState).button,
+      ref: useMergedRefs((state as DropdownState).button.ref, triggerRef as React.RefObject<HTMLButtonElement>),
+    };
+
+    triggerSlot = (state as DropdownState).button;
+  }
+
+  // the trigger should open/close the popup on click or blur
+  const { onBlur: onTriggerBlur, onClick: onTriggerClick, onKeyDown: onTriggerKeyDown } = triggerSlot;
+  triggerSlot.onBlur = (event: React.FocusEvent<HTMLButtonElement> & React.FocusEvent<HTMLInputElement>) => {
```

Comment:
You could use `useMergedEventCallbacks` for these, and skip defining/calling `onTriggerBlur`, etc.
```suggestion
  triggerSlot.onBlur = useMergedEventCallbacks(triggerSlot.onBlur, event => {
```